pkg/data/file.go

```diff
@@ -56,7 +66,55 @@ func convertURLToBytes(url string) (b []byte, contentType string, filename strin
 	if strings.HasPrefix(url, "data:") {
 		return convertDataURIToBytes(url)
 	}
+	if matches := minioURLPattern.FindStringSubmatch(url); matches != nil {
+		if len(matches) < 2 {
+			return nil, "", "", fmt.Errorf("invalid blob storage url: %s", url)
+		}
+		return fetchFileFromBlobStorage(uuid.FromStringOrNil(matches[1]))
+	}
+	return fetchExternalURL(url)
+}
+
+func fetchFileFromBlobStorage(urlUID uuid.UUID) (b []byte, contentType string, filename string, err error) {
```

Comment:
This is implemented as a pure function (without any receiver) but, underneath, it initialises (or leverages some initialised) clients and dependencies. This is misleading and I'd even say unidiomatic. A much more natural way / more aligned with our codebase would be having a `BlobFetcher` structure which holds the dependencies (artifact client, MinIO client, even an HTTP client). These should be initialised on the `main` package and injected to this struct.

Reply:
I suspect this issue is caused by the current data package implementation.

The `fetchFileFromBlobStorage` function is invoked by `NewFileFromURL()`, but `NewFileFromURL()` doesn’t inject any client since it might be used in various places.

@jvallesm, are you suggesting that we should inject the client through `NewFileFromURL()`? If so, should we also inject the Resty client from the outside? This would make the function signature something like:
`NewFileFromURL(url, fetcher)`.

Reply:
Is it like what I proposed a few days ago?
![image](https://github.com/user-attachments/assets/98739274-a60f-426a-b05f-05c480afe5f9)

Reply:
Another approach is we can make the BlobFetcher be a singleton.

Reply:
Actually I saw no clients for `NewFileFromURL` (will it be used?), but I do see clients for other functions that use the `convertURLToBytes`, so the question remains valid.

Instead of passing the clients as function params, I see much more natural creating a `BlobFetcher` or `BinaryFetcher` (to align with current naming, i.e., `NewBinaryFromURL`) that acts as the receiver of the `ConvertURLToBytes` (I wrote it as exported because we might want to place this in another package, not sure about that).

```go
type BinaryFetcher struct {
	HTTPCLient     *resty.Client
	ArtifactClient artifactpb.ArtifactPrivateServiceClient // By having this injected, main.go is responsible of closing the connection.
	MinIOClient    miniox.MinioI
}

func (f *BinaryFetcher) FetchFromURL(url string) (b []byte, contentType string, filename string, err error) {
	if strings.HasPrefix(url, "data:") {
		return w.convertDataURIToBytes(url)
	}
	if matches := minioURLPattern.FindStringSubmatch(url); matches != nil {
		if len(matches) < 2 {
			return nil, "", "", fmt.Errorf("invalid blob storage url: %s", url)
		}
		return w.fetchFileFromBlobStorage(uuid.FromStringOrNil(matches[1]))
	}
	return w.fetchExternalURL(url)
}
```
Then, from Pipeline we can do the following (e.g. for the [audio](https://github.com/instill-ai/pipeline-backend/blob/main/pkg/service/pipeline.go) or the [binary](https://github.com/instill-ai/pipeline-backend/blob/main/pkg/service/pipeline.go) fetchers):
```go
			case "audio", "audio/*":
				url := defaultValueMap[k].(string)
				if v != nil {
					url = v.GetStringValue()
				}
				data, err := s.binaryFetcher.FetchFromURL(url)
				if err != nil {
					return fmt.Errorf("fetching binary data from URL: %w", err)
				}

				variable[k], err = data.CreateAudioData(data)
				if err != nil {
					return fmt.Errorf("creating audio data: %w", err)
				}
			case "file", "*/*":
				url := defaultValueMap[k].(string)
				if v != nil {
					url = v.GetStringValue()
				}
				data, err := s.binaryFetcher.FetchFromURL(url)
				if err != nil {
					return fmt.Errorf("fetching binary data from URL: %w", err)
				}

				variable[k], err = data.NewBinaryFromBytes(data) // And we can remove NewBinaryFromURL
				if err != nil {
					return fmt.Errorf("creating audio data: %w", err)
				}

```

Reply:
I think introducing a BinaryFetcher is a good idea. However, there are two concerns:

1. We might need to create a file from a URL within components, which would require injecting the BlobFetcher into each component implementations.
2. We also create files by URL in the data marshalling functions, meaning the BlobFetcher would need to be injected there as well.

If we can make BinaryFetcher a global singleton and initialize it in main.go, I guess it could simplify things.

Reply:
My concern with the singleton is that it hides these dependencies. Especially with the 1st point, this would prevent having components as standalone executors (using the package to execute the task or a CLi like `instill component execute openai --task=TEXT_GENERATION`) as sometimes we've discussed. The dangers of singletons are well documented so I'm against them using them lightly (I reckon there are cases where they make sense).

Reply:
Well, to be a bit fair, it won't _prevent_ that use case but I imagine the dev experience being frustrating because you didn't know you needed to initialise that singleton.

Reply:
Yeah, make senses

Reply:
Then I guess we probably need to make data.Marshal and data.Unmarhsal to be something like this
```golang
marshaler := data.NewMarshaler().SetBinaryFetcher(binaryFetcher)
value, _ := marshaler.marshal(data)
```

Reply:
Looks good 👍 Also, we can always default to the HTTP fetcher, the artifact object conversion + minIO fetch is only an optimisation to save egress costs. That way, the use cases you mentioned would work without any extra plumbing.

Reply:
```
marshaler := data.NewMarshaler().SetBinaryFetcher(binaryFetcher)
value, _ := marshaler.marshal(data)
```

Not so sure what it means here.

Where do we init this marshaler?

Reply:
Since this won't be tackled in this PR, I responded in the ticket with a more detailed proposal.